lib/mapObjects/CGTownBuilding.cpp

```diff
@@ -284,5 +285,220 @@ void CTownBonus::applyBonuses(CGHeroInstance * h, const BonusList & bonuses) con
+}
+
+void CTownRewardableBuilding::newTurn(CRandomGenerator & rand) const
+{
+	if (configuration.resetParameters.period != 0 && cb->getDate(Date::DAY) > 1 && ((cb->getDate(Date::DAY)-1) % configuration.resetParameters.period) == 0)
+	{
+		if(configuration.resetParameters.rewards)
+		{
+			cb->setObjProperty(town->id, ObjProperty::REWARD_RANDOMIZE, indexOnTV);
+		}
+		if(configuration.resetParameters.visitors)
+		{
+			cb->setObjProperty(town->id, ObjProperty::REWARD_CLEARED, indexOnTV);
+			cb->setObjProperty(town->id, ObjProperty::STRUCTURE_CLEAR_VISITORS, indexOnTV);
+		}
+	}
+}
+
+void CTownRewardableBuilding::setProperty(ui8 what, ui32 val)
+{
+	switch (what)
+	{
+		case ObjProperty::VISITORS:
+			visitors.insert(ObjectInstanceID(val));
+			break;
+		case ObjProperty::STRUCTURE_CLEAR_VISITORS:
+			visitors.clear();
+			break;
+		case ObjProperty::REWARD_RANDOMIZE:
+			initObj(cb->gameState()->getRandomGenerator());
+			break;
+		case ObjProperty::REWARD_SELECT:
+			selectedReward = val;
+			break;
+		case ObjProperty::REWARD_CLEARED:
+			onceVisitableObjectCleared = val;
+			break;
+	}
+}
+
+void CTownRewardableBuilding::heroLevelUpDone(const CGHeroInstance *hero) const
+{
+	grantRewardAfterLevelup(cb, configuration.info.at(selectedReward), town, hero);
+}
+
+void CTownRewardableBuilding::blockingDialogAnswered(const CGHeroInstance *hero, ui32 answer) const
+{
+	if(visitors.find(hero->id) != visitors.end())
+		return; // query not for this building
+	
+	if(answer == 0)
+	{
+		cb->setObjProperty(town->id, ObjProperty::STRUCTURE_CLEAR_VISITORS, indexOnTV);
+		return; // player refused
+	}
+
+	if(answer > 0 && answer-1 < configuration.info.size())
+	{
+		auto list = getAvailableRewards(hero, Rewardable::EEventType::EVENT_FIRST_VISIT);
+		grantReward(list[answer - 1], hero);
+	}
+	else
+	{
+		throw std::runtime_error("Unhandled choice");
+	}
+}
+
+void CTownRewardableBuilding::grantReward(ui32 rewardID, const CGHeroInstance * hero) const
+{
+	town->addHeroToStructureVisitors(hero, indexOnTV);
+	
+	grantRewardBeforeLevelup(cb, configuration.info.at(rewardID), hero);
+	
+	// hero is not blocked by levelup dialog - grant remainer immediately
+	if(!cb->isVisitCoveredByAnotherQuery(town, hero))
+	{
+		grantRewardAfterLevelup(cb, configuration.info.at(rewardID), town, hero);
+	}
+}
+
+bool CTownRewardableBuilding::wasVisitedBefore(const CGHeroInstance * contextHero) const
+{
+	switch (configuration.visitMode)
+	{
+		case Rewardable::VISIT_UNLIMITED:
+			return false;
+		case Rewardable::VISIT_ONCE:
+			return onceVisitableObjectCleared;
+		case Rewardable::VISIT_PLAYER:
+			return false; //not supported
+		case Rewardable::VISIT_BONUS:
+			return contextHero->hasBonusFrom(Bonus::TOWN_STRUCTURE, Bonus::getSid32(town->town->faction->getIndex(), bID));
+		case Rewardable::VISIT_HERO:
+			return visitors.find(contextHero->id) != visitors.end();
+		default:
+			return false;
+	}
+}
+
+void CTownRewardableBuilding::onHeroVisit(const CGHeroInstance *h) const
```

Comment:
We might need some refactoring (either now or later) to minimize duplicated code in places like here.